src/ActionsLinux.cpp

```diff
@@ -0,0 +1,210 @@
+        if (dst_file_info.Exists())
+        {
+            if (dst_file_info.IsDirectory())
+            {
+                result.m_ReturnCode = -1;
+                snprintf(tmpBuffer, sizeof(tmpBuffer), "The target path %s already exists as a directory.", target_file);
+                break;
+            }
+
+            if (dst_file_info.IsReadOnly())
+            {
+                result.m_ReturnCode = -1;
+                snprintf(tmpBuffer, sizeof(tmpBuffer), "The target path %s already exists and is read-only.", target_file);
+                break;
+            }
+        }
+
+        if ((src_stat.st_mode & S_IFMT) == S_IFLNK)
+        {
+            // It's a symlink
+
+            // Add 2 bytes, 1 for the null character and one so we can tell that the readlink return value did not get truncated
+            size_t bufferSize = src_stat.st_size + 2; 
+            char* link_contents = static_cast<char*>(LinearAllocate(&scratch, bufferSize, 1));
+            int link_size = readlink(src_file, link_contents, bufferSize);
+            if (link_size == bufferSize)
+            {
+                result.m_ReturnCode = -1;
+                snprintf(tmpBuffer, sizeof(tmpBuffer), "The source symlink %s could not be read with a consistent size.", src_file);
+                break;
+            }
+
+            if (dst_file_info.Exists())
+                unlink(target_file);
+
+            if (symlink(link_contents, target_file) != 0)
+            {
+                result.m_ReturnCode = -1;
+                snprintf(tmpBuffer, sizeof(tmpBuffer), "The target symlink %s could not be created.", target_file);
+                break;
+            }
+
+            // Mark the stat cache dirty
+            StatCacheMarkDirty(stat_cache, target_file, target_files[i].m_FilenameHash);
+        }
+        else
+        {
+            // It's a regular file
+            in_file = open(src_file, O_RDONLY);
+            if (in_file == -1)
+            {
+                result.m_ReturnCode = -1;
+                snprintf(tmpBuffer, sizeof(tmpBuffer), "The source file %s could not be opened for reading: %s", src_file, strerror(errno));
+                break;
+            }
+
+            // Ensure that the target file is opened with a writable mode, even if the input file was readonly
+            out_file = open(target_file, O_WRONLY | O_CREAT | O_TRUNC, src_stat.st_mode | S_IWUSR);
+            if (out_file == -1)
+            {
+                result.m_ReturnCode = -1;
+                snprintf(tmpBuffer, sizeof(tmpBuffer), "The destination file %s could not be opened for writing: %s", src_file, strerror(errno));
+                break;
+            }
+
+            do {
+#ifdef FICLONE
+                // See if it is on the same filesystem (which it usually will be) - if it is, we can try using the FICLONE ioctl to do a lightweight copy-on-write copy
+                struct stat dst_stat;
+                if (fstat(out_file, &dst_stat) != 0)
```

Comment:
Ah, yeah, ok, I guess that if I am just falling back in case of errors no matter what then we can just let it fail like that. I originally had this because I wanted to try and provide better error handling if the ioctl failed, but that was before I'd set it up to silently failover to splicing.

Reply:
Yeah I like the silent-failover, since these aren't "errors" per se.